src/hypothesis/internal/conjecture/engine.py

```diff
@@ -378,51 +378,48 @@ def _run(self):
         while self.changed > change_counter:
             change_counter = self.changed
 
-            self.debug('Random interval deletes')
-            failed_deletes = 0
-            while self.last_data.intervals and failed_deletes < 10:
-                if self.random.randint(0, 1):
-                    u, v = self.random.choice(self.last_data.intervals)
-                else:
-                    n = len(self.last_data.buffer) - 1
-                    u, v = sorted((
-                        self.random.choice(self.last_data.intervals)
-                    ))
-                if (
-                    v < len(self.last_data.buffer)
-                ) and self.incorporate_new_buffer(
-                    self.last_data.buffer[:u] +
-                    self.last_data.buffer[v:]
-                ):
-                    failed_deletes = 0
-                else:
-                    failed_deletes += 1
-
             self.debug('Structured interval deletes')
-            i = 0
-            while i < len(self.last_data.intervals):
-                u, v = self.last_data.intervals[i]
-                if not self.incorporate_new_buffer(
-                    self.last_data.buffer[:u] +
-                    self.last_data.buffer[v:]
-                ):
-                    i += 1
 
-            if change_counter != self.changed:
-                self.debug('Restarting')
-                continue
+            k = len(self.last_data.intervals) // 2
+            while k > 0:
+                i = 0
+                while i + k <= len(self.last_data.intervals):
+                    bitmask = [True] * len(self.last_data.buffer)
 
-            self.debug('Lexicographical minimization of whole buffer')
-            minimize(
-                self.last_data.buffer, self.incorporate_new_buffer,
-                cautious=True
-            )
+                    for u, v in self.last_data.intervals[i:i + k]:
+                        for t in range(u, v):
+                            bitmask[t] = False
+
+                    u, v = self.last_data.intervals[i]
+                    if not self.incorporate_new_buffer(hbytes(
+                        b for b, v in zip(self.last_data.buffer, bitmask)
+                        if v
+                    )):
+                        i += k
+                k //= 2
 
             if change_counter != self.changed:
                 self.debug('Restarting')
                 continue
 
-            self.debug('Replacing blocks with simpler blocks')
+            self.debug('Bulk replacing blocks with simpler blocks')
+            i = 0
+            while i < len(self.last_data.blocks):
```

Comment:
Yup.

Note that we don't mutate the list, we replace it with an entirely new one.